### ISSUE TITLE
Fisher double shift eigendecomp doesn't always work

### ISSUE THREAD

**Issue by AtheMathmo** [AtheMathmo/rusty-machine#64]
## What is the issue?

The current implementation disagrees with other standard libraries like numpy, wolframalpha, etc.

The following example is broken:

```
[1, 22, 27]
[22, 29, 36]
[27, 36, 45]
```

_Edit: The matrix should be:_
```
[17, 22, 27]
[22, 29, 36]
[27, 36, 45]
```

The true eigenvalues are: [90.4207, 0.5973, 0]
The rusty-machine eigenvalues are: [90.0834, 0.9086, -0.0349]

@zackmdavis has been investigating this issue. There is some useful discussion here: #46 

### First Steps

1. Modify the current implementation to use `MatrixSliceMut` instead of manually slicing.
2. Research other options for implementation.

**Comment by AtheMathmo** [comment]
This issue seems to be resolved by #98 . However - the matrix given in this issue does not have the suggested eigenvalues. This merits some further investigation but is at least somewhat positive!

**Comment by zackmdavis** [comment]
> the matrix given in this issue does not have the suggested eigenvalues

You mistyped it and I didn't bother writing a correction comment at the time (I might not be remembering correctly that the first entry was 11, but I'm pretty sure it wasn't 1).

**Comment by AtheMathmo** [comment]
Ha ok :D I'll have to recheck it. Thanks 

---- Zack M. Davis wrote ----

**Comment by AtheMathmo** [comment]
After checking with the actual matrix it looks like we still have some issues.. I'd suggest it is best to merge in the cleanup of decomposition in #98 and then tackle this bug afterwards. It may require some heavy work to the eigendecomp algorithms.

**Comment by AtheMathmo** [comment]
@zackmdavis: I'm pretty sure that I found the (read _a_) bug!

The application of the givens rotation on the right was acting on too small a block of the matrix. I'll add this fix to the active decomposition PR (#98) .
